docs/Internals.md

```diff
@@ -0,0 +1,124 @@
+# Common APIs and Algorithm Internals
+
+## Common APIs
+
+Below are the primary public-facing interfaces of Asherah.
+
+**NOTE:** The interfaces below are from the Java implementation of the SDK.
+
+### Primary SDK Interfaces
+
+The below interfaces implement the session factory using the step builder pattern.
+
+```java
+class AppEncryptionSessionFactory {
+  static MetastoreStep newBuilder(String productId, String systemId);
+  
+  AppEncryption<JSONObject, byte[]> getAppEncryptionJson(String partitionId);
+  AppEncryption<byte[], byte[]> getAppEncryptionBytes(String partitionId);
+  AppEncryption<JSONObject, JSONObject> getAppEncryptionJsonAsJson(String partitionId);
+  AppEncryption<byte[], JSONObject> getAppEncryptionBytesAsJson(String partitionId);
+  
+  void close();
+}
+ 
+interface MetastoreStep {
+  CryptoPolicyStep withMetastorePersistence(MetastorePersistence<JSONObject> persistence);
+}
+
+interface CryptoPolicyStep {
+  KeyManagementServiceStep withCryptoPolicy(CryptoPolicy policy);
+}
+
+interface KeyManagementServiceStep {
+  BuildStep withKeyManagementService(KeyManagementService keyManagementService);
+}
+
+interface BuildStep {
+  BuildStep withMetricsEnabled();
+
+  AppEncryptionSessionFactory build();
+}
+```
+
+Cryptographic operations are performed using the methods provided in the AppEncryption interface.
+
+```java
+// <P> The payload type being encrypted
+// <D> The Data Row Record type
+interface AppEncryption<P, D> {
+  P decrypt(D dataRowRecord);
+  D encrypt(P payload);
+
+  Optional<P> load(String persistenceKey, Persistence<D> dataPersistence);
+  String store(P payload, Persistence<D> dataPersistence);
+  void store(String key, P payload, Persistence<D> dataPersistence);
+
+  void close();
+}
+```
+  
+For the load/store usage model, we also need to implement the Persistence interface
```

Comment:
can optionally add a link to `../README.md#store--load` for the "load/store" text